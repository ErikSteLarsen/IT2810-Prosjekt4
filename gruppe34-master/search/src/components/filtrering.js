import React, { Component } from "react";
import { withApollo } from "react-apollo";
import { compose } from "redux";
import { connect } from "react-redux";
import "../stylesheets/filtrering.css";
import Card from "@material-ui/core/Card";
import CardContent from "@material-ui/core/CardContent";
import { changeSorting } from "../actions/changeSorting";
import TextField from "@material-ui/core/TextField";
import { fromFilter } from "../actions/fromFilter";
import { toFilter } from "../actions/toFilter";
import Button from "@material-ui/core/Button";

/**
 *
 * @author Erik
 */

class Filtrering extends Component {

  /* Når man endrer valget av sortering gjør denne funksjonen at det også
     endres i store.   */
  onSelectChange = event => {
    this.props.dispatch(changeSorting(event.target.value));
  };

  /* Når man endrer verdi på "fra"-årstall så sørger denne funksjonen for at
     store også endres.   */
  onFromChange = event => {
    let value = Number(event.target.value) - 1;
    this.props.dispatch(fromFilter(value));
  };

  /* Når man endrer verdi på "til"-årstall så sørger denne funksjonen for at
     store også endres.   */
  onToChange = event => {
    let value = Number(event.target.value) + 1;
    this.props.dispatch(toFilter(value));
  };
  onChangePress = event => {
    this.props.fetchFunction(event);
  };

  render() {
    // Dette er konstanter for å style Material-UI komponenter vi har brukt.
    // pga vanskelig å style de uten å direkte legge in style={puttInnKonstantHer}.
    const numberInput1 = {
      width: "100px",
      marginRight: "15px",
      marginLeft: "15px",
    };
    const numberInput2 = {
      width: "100px",
      marginRight: "15px",
      marginLeft: "15px",
    };

    const style1 = {
      backgroundColor: "#fafafa",
      height: "auto",
    };
    const updateButton = {
      width: "90px",
      height: "40px",
      backgroundColor: "#f1f1f1",
      border: "1px solid",
      borderColor: "grey",
      marginRight: "15px",
      marginLeft: "15px",
      borderRadius: "10px"
    };
    const styleOptions = {
      marginRight: "15px",
      marginLeft: "15px",
    }
    return (
      <Card style={style1} className="infoBox">
        <CardContent className="cardClass">
          <div className="filteringContainer">
            <TextField
              id="standard-select-currency-native"
              select
              label="Select sorting"
              onChange={this.onSelectChange}
              SelectProps={{
                native: true
              }}
              margin="normal"
              style={styleOptions}
            >
              <option value="Default_sorting">Sort by ranking</option>
              <option value="Title_ASC">Title ASC</option>
              <option value="Title_DESC">Title DESC</option>
              <option value="Year_ASC">Year ASC</option>
              <option value="Year_DESC">Year DESC</option>
            </TextField>

            <TextField
              style={numberInput1}
              id="standard-number"
              label="From year:"
              type="number"
              InputLabelProps={{
                shrink: true
              }}
              placeholder="1900"
              onChange={this.onFromChange}
              margin="normal"
              data-cy="sort_from_year"
            />

            <TextField
              style={numberInput2}
              id="standard-number2"
              label="To year:"
              type="number"
              placeholder="2018"
              InputLabelProps={{
                shrink: true
              }}
              onChange={this.onToChange}
              margin="normal"
              data-cy="sort_to_year"
            />

            <Button
              style={updateButton}
              onClick={event => this.onChangePress(event)}
              data-cy="update_filter_btn"
            >
              UPDATE
            </Button>
          </div>
        </CardContent>
      </Card>
    );
  }
}

// For å få tilgang til store med connect lenger ned.
const mapStateToProps = state => {
  return {
    moviesStore: state.moviesStore,
    text: state.changeSearch,
    clickedMovie: state.clickedMovie,
    clickedMovieTitle: state.clickedMovieTitle,
    sortBy: state.sortBy,
    fromYear: state.fromFilter,
    toYear: state.toFilter
  };
};

//Wrapping av komponent med begge providerene våres. (Apollo og Redux).
export default compose(
  withApollo,
  connect(mapStateToProps)
)(Filtrering);
