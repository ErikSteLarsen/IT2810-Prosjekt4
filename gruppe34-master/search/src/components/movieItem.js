import React, { Component } from "react";
import { withApollo, graphql } from "react-apollo";
import { compose } from "redux";
import { connect } from "react-redux";
import "../stylesheets/movieItem.css";
import Card from "@material-ui/core/Card";
import CardContent from "@material-ui/core/CardContent";
import Typography from "@material-ui/core/Typography";
import Modal from "@material-ui/core/Modal";
import { addClickedMovie } from "../actions/addClickedMovie";
import { getSpecifiedTitleQuery, updateRanking } from "../queries/queries.js";

/**
 *
 * @author Erik
 * Eksporterer også en ukoblet komponent for testing
 */

export class MovieItem extends Component {
  /*
   *  Bruker vanlig statehandling i redux ettersom dette er veldig enkel
   *  håndtering av noen få lokale variabler.
   *  State-handling av "MODAL"-komponenten vår.
   *  Mer om dette er forklart i README.MD.
   */
  constructor(props) {
    super(props);
    this.state = {
      open: false,
      inputRating: ""
    };
  }

  /* Denne funksjonen laster inn mer data om det spesifikke elementet
   * du trykker på og viser det i en MODAL-komponent.
   */
  handleOpen = data => {
    this.props.client
      .query({
        query: getSpecifiedTitleQuery,
        variables: {
          Title: data
        }
      })
      .then(res => {
        this.props.dispatch(addClickedMovie(res.data.title[0]));
        this.setState({ open: true });
      });
  };

  /* Denne funksjonen lukker MODAL-komponenten.
   */
  handleClose = () => {
    // Bruker vanlig statehandling i redux ettersom dette er veldig enkel
    // håndtering av state og bare er mer jobb å fikse i redux.
    // State-handling av "MODAL"-komponenten vår.
    this.setState({ open: false });
  };

  /* Erstatter tittelen en liten endring dersom den er for lang.  */
  fillTitle = title => {
    if (title.length > 40) {
      return title.substring(0, 40) + "...";
    } else {
      return title;
    }
  };

  /* Denne funksjonen sørger for at man ikke kan putte inn
     feil values når man skal skrive inn en rating   */
  setInputValue = value => {
    let re = /[0-9]*\.?[0-9]+/g;
    let valueInsert = "";
    if (re.test(value.toString()) && value >= 0 && value <= 10) {
      valueInsert = value;
    }
    this.setState({ inputRating: valueInsert });
  };

  /* Denne funksjonen gjører en mutation opp mot databasen
     som sørger for en endring når man legger til en rating
     på en film. Her regnes det ut gjennomsnittlig rating osv.*/
  addRatingToMovie = () => {
    let ntnuVotes = this.props.clickedMovie.ntnuVotes;
    let ntnuRating = this.props.clickedMovie.ntnuRating;
    let userRating = this.state.inputRating.toString();

    if (ntnuRating === null) {
      ntnuRating = "0";
    }

    if (userRating !== "") {
      this.props.client
        .mutate({
          mutation: updateRanking,
          variables: {
            Title: this.props.clickedMovie.Title,
            ntnuVotes: ntnuVotes,
            ntnuRating: ntnuRating,
            userRating: userRating
          }
        })
        .then(res =>
          this.props.dispatch(addClickedMovie(res.data.updateRanking))
        );
    } else {
      alert("Please insert a correct value.");
    }
  };

  // Render function.
  render() {
    const marginBottomLast = {
      marginBottom: "5px",
      borderBottom: "1px solid",
      width: "100%"
    };
    const setText = {
      fontWeight: "bold"
    };
    const titleStyle = {
      fontWeight: "bold",
      marginBottom: "20px"
    };
    const boldText2 = {
      fontWeight: "bold"
    };
    const year = {
      fontSize: "13px"
    };
    const movieItem = {
      backgroundColor: "rgb(250, 250, 250)",
      cursor: "pointer",
      userSelect: "none"
    };
    return (
      <React.Fragment>
        <Card
          style={movieItem}
          className="movieItem"
          id={this.props.title}
          onClick={data => this.handleOpen(this.props.title)}
          data-cy="movie_item"
        >
          <CardContent data-cy="card">
            <Typography style={boldText2} data-cy="title">
              {this.fillTitle(this.props.title)}
            </Typography>
            <Typography style={year} variant="subtitle1" color="textSecondary">
              {this.props.year}
            </Typography>
          </CardContent>
        </Card>

        <Modal open={this.state.open} onClose={this.handleClose}>
          <div className="modal" data-cy="result_modal">
            <Typography style={titleStyle} variant="h5">
              {this.props.clickedMovie.Title}
            </Typography>
            <Typography style={boldText2} variant="subtitle1">
              Released:
            </Typography>
            <Typography style={marginBottomLast} variant="subtitle1">
              {this.props.clickedMovie.Released}
            </Typography>
            <Typography style={boldText2} variant="subtitle1">
              Runtime:
            </Typography>
            <Typography style={marginBottomLast} variant="subtitle1">
              {this.props.clickedMovie.Runtime}
            </Typography>
            <Typography style={boldText2} variant="subtitle1">
              Genre:
            </Typography>
            <Typography style={marginBottomLast} variant="subtitle1">
              {this.props.clickedMovie.Genre}
            </Typography>
            <Typography style={boldText2} variant="subtitle1">
              Director:
            </Typography>
            <Typography style={marginBottomLast} variant="subtitle1">
              {this.props.clickedMovie.Director}
            </Typography>
            <Typography style={boldText2} variant="subtitle1">
              Awards:
            </Typography>
            <Typography style={marginBottomLast} variant="subtitle1">
              {this.props.clickedMovie.Awards}
            </Typography>
            <Typography style={boldText2} variant="subtitle1">
              Average NTNU Rating:
            </Typography>
            <Typography style={marginBottomLast} variant="subtitle1">
              {Math.round(this.props.clickedMovie.ntnuRating * 100) / 100}
            </Typography>

            <button
              className="closeButton"
              onClick={this.handleClose}
              data-cy="button_close"
            >
              X
            </button>

            <Typography style={setText} variant="subtitle1">
              Set rating for this movie (0-10):
            </Typography>
            <input
              placeholder="e.g. 9,4"
              onChange={event => this.setInputValue(event.target.value)}
              value={this.state.inputRating}
              max="10"
              min="0"
              className="inputRating"
              type="number"
              data-cy="ntnu_rating_input"
            />
            <button
              type="submit"
              className="buttonRating"
              onClick={this.addRatingToMovie}
            >
              Set Rating
            </button>
          </div>
        </Modal>
      </React.Fragment>
    );
  }
}

// For å få tilgang til store med connect lenger ned.
const mapStateToProps = state => {
  return {
    moviesStore: state.moviesStore,
    text: state.changeSearch,
    clickedMovie: state.clickedMovie
  };
};

//Wrapping av komponent med begge providerene våres. (Apollo og Redux).
export default compose(
  graphql(updateRanking, {
    name: "updateRanking",
    options: { refetchQueries: ["getSpecifiedTitleQuery"] }
  }),
  withApollo,
  connect(mapStateToProps)
)(MovieItem);
