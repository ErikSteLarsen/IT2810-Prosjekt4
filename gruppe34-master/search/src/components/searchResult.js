import React, { Component } from "react";
import { withApollo } from "react-apollo";
import { compose } from "redux";
import { connect } from "react-redux";
import "../stylesheets/searchResult.css";
import MovieItem from "./movieItem.js";
/*
import {addClickedMovieTitle} from '../actions/addClickedMovieTitle';
import {getAllTitlesQuery} from '../queries/queries.js';
*/

/**
 * Takes data from mobx and shows it in table
 * @author Magnus and Erik
 */

/*
This component is the component that the search result appears in.
Its a very simple component, the main thing it does is map every object fetched from the server
to a MovieItem component.
The function that transforms data is explained below.
*/

class SearchResult extends Component {

  /*
  This function transforms the array of objects in the store,
  to a 2D array cotaining Title and Year for each movie in the result.
  */
  transformData = () => {
    let data = [];
    if (this.props.moviesStore.title) {
      this.props.moviesStore.title.forEach(movie => {
        data.push([movie.Title, movie.Year]);
      });
    }
    return data;
  };

  render() {
  let data = this.transformData();
    return (
      <div className="searchResult" data-cy="search_results">
        {data.map(movie => (
          <MovieItem title={movie[0]} year={movie[1]} key={movie[0]} />
        ))}
      </div>
    );
  }
}

const mapStateToProps = state => {
  return {
    moviesStore: state.moviesStore,
  }
};

export default compose(
  withApollo,
  connect(mapStateToProps)
)(SearchResult);
