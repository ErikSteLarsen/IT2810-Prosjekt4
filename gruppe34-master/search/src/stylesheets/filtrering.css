/**
 *
 * @author Erik
 */

.infoBox {
  min-width: 200px;
  max-width: 700px;
  width: 50%;
  height: 100px;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
}

.sortBy {
  margin-right: 1%;
}

.cardClass {
  width: 100%;
  margin-top: 10px;
  display: flex;
  flex-direction: row;
  justify-content: center;
}

.filterdiv {
  margin-left: 10%;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
}

.filterboxes {
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: center;
}

.filterstyle {
  margin-left: 5%;
  margin-right: 1%;
}

.filteringContainer {
  width: 100%;
  height: auto;
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  align-items: center;
  justify-content: center;
}
