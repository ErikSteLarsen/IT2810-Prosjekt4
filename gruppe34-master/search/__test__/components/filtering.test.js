import React from "react";
import { MockedProvider } from "react-apollo/test-utils";
import { Filtrering } from "../../src/components/filtrering";
import renderer from "react-test-renderer";
import configureStore from "redux-mock-store";

const middlewares = [];
const mockStore = configureStore(middlewares);

const initialState = [{}];
const store = mockStore(initialState);

const mocks = [];

it("renders without error", () => {
  const tree = renderer
    .create(
      <MockedProvider mocks={mocks}>
        <Filtrering store={store} />
      </MockedProvider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
