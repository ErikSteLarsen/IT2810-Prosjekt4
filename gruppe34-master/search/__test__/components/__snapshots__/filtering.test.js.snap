// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without error 1`] = `
<div
  className="MuiPaper-root-2 MuiPaper-elevation1-5 MuiPaper-rounded-3 MuiCard-root-1 infoBox"
  style={
    Object {
      "backgroundColor": "#fafafa",
      "height": "auto",
    }
  }
>
  <div
    className="MuiCardContent-root-29 cardClass"
  >
    <div
      className="filteringContainer"
    >
      <div
        className="MuiFormControl-root-30 MuiFormControl-marginNormal-31"
        style={
          Object {
            "marginLeft": "15px",
            "marginRight": "15px",
          }
        }
      >
        <label
          className="MuiFormLabel-root-41 MuiInputLabel-root-34 MuiInputLabel-formControl-35 MuiInputLabel-animated-38"
          data-shrink={false}
          htmlFor="standard-select-currency-native"
        >
          Select sorting
        </label>
        <div
          className="MuiInputBase-root-68 MuiInput-root-55 MuiInput-underline-59 MuiInputBase-formControl-69 MuiInput-formControl-56"
          onClick={[Function]}
        >
          <div
            className="MuiSelect-root-48"
          >
            <select
              aria-invalid={false}
              className="MuiSelect-select-49 MuiInputBase-input-78 MuiInput-input-63"
              disabled={false}
              id="standard-select-currency-native"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              required={false}
            >
              <option
                value="Default_sorting"
              >
                Sort by ranking
              </option>
              <option
                value="Title_ASC"
              >
                Title ASC
              </option>
              <option
                value="Title_DESC"
              >
                Title DESC
              </option>
              <option
                value="Year_ASC"
              >
                Year ASC
              </option>
              <option
                value="Year_DESC"
              >
                Year DESC
              </option>
            </select>
            <svg
              aria-hidden="true"
              className="MuiSvgIcon-root-85 MuiSelect-icon-54"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
          </div>
        </div>
      </div>
      <div
        className="MuiFormControl-root-30 MuiFormControl-marginNormal-31"
        data-cy="sort_from_year"
        style={
          Object {
            "marginLeft": "15px",
            "marginRight": "15px",
            "width": "100px",
          }
        }
      >
        <label
          className="MuiFormLabel-root-41 MuiInputLabel-root-34 MuiInputLabel-formControl-35 MuiInputLabel-animated-38 MuiInputLabel-shrink-37"
          data-shrink={true}
          htmlFor="standard-number"
        >
          From year:
        </label>
        <div
          className="MuiInputBase-root-68 MuiInput-root-55 MuiInput-underline-59 MuiInputBase-formControl-69 MuiInput-formControl-56"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            className="MuiInputBase-input-78 MuiInput-input-63 MuiInputBase-inputType-81 MuiInput-inputType-66"
            disabled={false}
            id="standard-number"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            placeholder="1900"
            required={false}
            type="number"
          />
        </div>
      </div>
      <div
        className="MuiFormControl-root-30 MuiFormControl-marginNormal-31"
        data-cy="sort_to_year"
        style={
          Object {
            "marginLeft": "15px",
            "marginRight": "15px",
            "width": "100px",
          }
        }
      >
        <label
          className="MuiFormLabel-root-41 MuiInputLabel-root-34 MuiInputLabel-formControl-35 MuiInputLabel-animated-38 MuiInputLabel-shrink-37"
          data-shrink={true}
          htmlFor="standard-number2"
        >
          To year:
        </label>
        <div
          className="MuiInputBase-root-68 MuiInput-root-55 MuiInput-underline-59 MuiInputBase-formControl-69 MuiInput-formControl-56"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            className="MuiInputBase-input-78 MuiInput-input-63 MuiInputBase-inputType-81 MuiInput-inputType-66"
            disabled={false}
            id="standard-number2"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            placeholder="2018"
            required={false}
            type="number"
          />
        </div>
      </div>
      <button
        className="MuiButtonBase-root-120 MuiButton-root-94 MuiButton-text-96 MuiButton-flat-99"
        data-cy="update_filter_btn"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "backgroundColor": "#f1f1f1",
            "border": "1px solid",
            "borderColor": "grey",
            "borderRadius": "10px",
            "height": "40px",
            "marginLeft": "15px",
            "marginRight": "15px",
            "width": "90px",
          }
        }
        tabIndex="0"
        type="button"
      >
        <span
          className="MuiButton-label-95"
        >
          UPDATE
        </span>
        <span
          className="MuiTouchRipple-root-123"
        />
      </button>
    </div>
  </div>
</div>
`;
